File: src/main/java/eu/qwsome/sql/SubselectValueHolder.java
package eu.qwsome.sql;

import eu.qwsome.sql.condition.ValueHolder;

/**
 * This class represents a subselect value in sql.
 *
 * @author Lukáš Kvídera
 */
public class SubselectValueHolder implements ValueHolder {

  private final Select.ConditionsBuiltPhase value;


  private SubselectValueHolder(final Select.ConditionsBuiltPhase value) {
    this.value = value;
  }


  public static SubselectValueHolder subselect(final Select.ConditionsBuiltPhase value) {
    return new SubselectValueHolder(value);
  }


  @Override
  public String getSql() {
    return '(' + this.value.toSql() + ')';
  }


  @Override
  public Object getValue() {
    return this.value.toValues();
  }

  @Override
  public ValueHolder apply(final String functionName) {
    return new FunctionWrappedHolder(functionName, this);
  }
}

File: src/main/kotlin/eu/qwsome/sql/SubselectValueHolder.kt
/*
 * This class represents a subselect value in sql.
 *
 * @author Lukáš Kvídera
 */
package eu.qwsome.sql

import eu.qwsome.sql.condition.ValueHolder
import eu.qwsome.sql.condition.withFunction

class SubselectValueHolder private constructor(private val value: Select.ConditionsBuiltPhase) : ValueHolder {

    override fun getSql(): String {
        return "(${value.toSql()})"
    }

    override fun getValue(): Any? {
        return value.toValues()
    }

    override fun apply(functionName: String): ValueHolder = withFunction(functionName)

    companion object {
        @JvmStatic
        fun subselect(value: Select.ConditionsBuiltPhase): SubselectValueHolder {
            return SubselectValueHolder(value)
        }
    }
}
